cgen error on `println( []?int{len: 3, init: none} )`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.5 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0

getwd: /v/vnew
vmodules: /home/delian/.vmodules
vroot: /v/vnew
vexe: /v/vnew/v
vexe mtime: 2023-03-07 08:10:13
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 1a48d08.784592a

Git version: git version 2.39.0
Git vroot status: weekly.2023.10-2-g784592af-dirty (2 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg another_array_of_options_bug.v`
```v
println( []?int{len: 3, init: none} )

```

**What did you expect to see?**

A compiled program, that can also be run, without crashing.


**What did you see instead?**
```
==================
/tmp/v_1000/another_array_of_options_bug.12902855345803147367.tmp.c:12316: error: cannot convert 'extern struct IError' to 'unsigned char'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -353,7 +353,11 @@ fn (mut g Gen) array_init_with_fields(node ast.ArrayInit, elem_type Type, is_amp
 		g.write(' (voidptr)${tmp})')
 	} else if node.has_default {
 		g.write('&(${elem_styp}[]){')
-		g.expr_with_cast(node.default_expr, node.default_type, node.elem_type)
+		if node.elem_type.has_flag(.option) {
+			g.expr_with_opt(node.default_expr, node.default_type, node.elem_type)
+		} else {
+			g.expr_with_cast(node.default_expr, node.default_type, node.elem_type)
+		}
 		g.write('})')
 	} else if node.has_len && node.elem_type == ast.string_type {
 		g.write('&(${elem_styp}[]){')

--- a/vlib/v/tests/option_array_init_test.v
+++ b/vlib/v/tests/option_array_init_test.v
@@ -1,0 +1,10 @@
+fn test_main() {
+	a := []?int{len: 3, init: none}
+	mut t := a[0]
+	assert t == none
+	t = a[1]
+	assert t == none
+	t = a[2]
+	assert t == none
+	assert a.len == 3
+}

